gadget/device_linux.go

```diff
@@ -202,17 +204,46 @@ func findParentDeviceWithWritableFallback() (string, error) {
 	if err != nil {
 		return "", err
 	}
-	return ParentDiskFromPartition(partitionWritable)
+	return ParentDiskFromMountSource(partitionWritable)
 }
 
-// ParentDiskFromPartition will find the parent disk device for the
-// given partition. E.g. /dev/nvmen0n1p5 -> /dev/nvme0n1
+// ParentDiskFromMountSource will find the parent disk device for the given
+// partition. E.g. /dev/nvmen0n1p5 -> /dev/nvme0n1.
 //
-// Note that this does not work for anything using device mapper (like
-// LUKS/LVM) yet.
-func ParentDiskFromPartition(partition string) (string, error) {
+// When the mount source is a symlink, it is resolved to the actual device that
+// is mounted. Should the device be one created by device mapper, it is followed
+// up to the actual underlying block device. As an example, this is how devices
+// are followed with a /writable mounted from an encrypted volume:
+//
+// /dev/mapper/ubuntu-data-<uuid> (a symlink)
+//   ⤷ /dev/dm-0 (set up by device mapper)
+//       ⤷ /dev/hda4 (returned by this function)
```

Comment:
One last point, should this be /dev/hda given that the last step goes up in the hierarchy to find the parent disk?

Reply:
Yes, good point!